index.js

```diff
@@ -46,32 +48,72 @@ function Filter(inputTree, options) {
   }
 
   this._cache = new Cache();
-  this._canProcessCache = Object.create(null);
-  this._destFilePathCache = Object.create(null);
+
+
+  //newly added
+  this._in = this._out = undefined;
 }
 
+//newly added
+Object.defineProperty(Filter.prototype, 'in', {
+  get() {
+    if (this._in) {
+      return this._in;
+    }
+    this._in = (new FSTree({
+      root: this.inputPaths[0],
+    }));
+    return this._in;
+  }
+});
+
+// should be moved to broccoli-plugin
+Object.defineProperty(Filter.prototype, 'out', {
+  get() {
+    if (this._out) {
+      return this._out;
+    }
+    this._out = (new FSTree({
+      root: this.outputPath
+    }));
+    return this._out;
+  }
+});
+
+
 Filter.prototype.build = function build() {
   var self = this;
   var srcDir = this.inputPaths[0];
   var destDir = this.outputPath;
   var paths = walkSync(srcDir);
 
-  this._cache.deleteExcept(paths).forEach(function(key) {
-    fs.unlinkSync(this.cachePath + '/' + key);
+  this._in = undefined;
+  this._out = undefined;
+
+  this.out.start();
+  this.in.start();
+
+  this._cache.deleteExcept(paths).forEach(function (key) {
+   fs.unlinkSync(this.cachePath + '/' + key);
   }, this);
 
+  var self = this;
+
   return mapSeries(paths, function rebuildEntry(relativePath) {
-    var destPath = destDir + '/' + relativePath;
     if (relativePath.slice(-1) === '/') {
-      mkdirp.sync(destPath);
+      self._out.mkdirpSync(relativePath);
     } else {
       if (self.canProcessFile(relativePath)) {
         return self.processAndCacheFile(srcDir, destDir, relativePath);
       } else {
         var srcPath = srcDir + '/' + relativePath;
-        symlinkOrCopySync(srcPath, destPath);
+        self._out.symlinkSync(srcPath, relativePath);
```

Comment:
here we should be able to do `this.in[0].resolvePath(relativePath)` instead of `srcPath`